### ISSUE TITLE
Многотенантная система

### ISSUE THREAD

**Issue by bestann** [prog-it/Asterisk-CDR-Viewer-Mod#43]
Хочу поставить вашу систему статистики, понравилась по виду и описанию.
Есть ли возможность разделять статистику по предприятиям со своими логинами и паролями?
Хотелось бы добавлять не каждый номер к предприятию, а определять соответствие контекстов и предприятий (2-4 контекста на 1 предприятие, т.к. в предприятии номера разделены по контекстам в зависимости от прав выхода на городские линии). Если в предприятии 100 номеров, удобнее конечно контекстом определить принадлежность предприятию.
Как вариант, префиксы номеров начинающиеся на определенные цифры.принадлежат такому-то предприятию.
Можно такое сделать или может уже сделано?

**Comment by prog-it** [comment]
Для разделения можно использовать "пользовательский" конфиг, в котором сможете задать разные настройки. Кнопку удаления записей можно отключить в конфиге.
Поиск с учетом Контекста отсутствует, может быть добавлю в следующем обновлении.
При поиске можно использовать Регулярные выражения в формате Asterisk. Например: Номер назначения - `_X95.`
Мне кажется, вам будет удобнее какая-нибудь CRM, например Freepbx

**Comment by bestann** [comment]
Во FreePBX нельзя сделать многотенантную систему. Во всяком случае в старых версиях.

**Comment by Korgoth1234** [comment]
Мне кажется нужно добавить еще одно поле в mysql с названием пользователя/предприятия и чтобы по нему CDR-Viever отображал информацию.
А вебку для разных предприятий как раз удобно развести через пользовательский конфиг.

**Comment by prog-it** [comment]
Название предприятия в этом случае в базу будет писать сам Asterisk ?
Если такой вариант подходит, то будет добавлено в ближайшее время. Название предприятия для поиска в базе прописано в конфиге, разграничение же по предприятиям по "пользовательскому конфигу".

**Comment by bestann** [comment]
Пока не знаю как реализовать грамотнее, я базы данных еще не знаю.

 Вариант 1 (Астер сам пишет в базу): в Астере в AstDB имени контекста, который присвоен абоненту, сопоставить имя предприятия. У нескольких контекстов одно и то же имя предприятия (т.к. телефоны с разными правами). Далее во время звонка имя, соответствующее контексту, вытаскиваем из AstDB, и помещаем значение в дополнительное поле базы cdr (аналогично имени файла записи разговора). На общей странице видны все звонки, на странице предприятия только звонки этого предприятия.
Но этот вариант добавляет дополнительный шаг в план нумерации и больше в базу будет писать в итоге.

Вариант 2. В базе данных MySQL/MariaDB составляется дополнительная таблица:

Рога и копыта - контексты roga-gorod, roga-internal, roga-mobile, roga-international
Петров и Ко - контексты petrov-gorod, petr-internal.....
И по соответствию подставляется имя предприятия.

У меня в контекстах всегда будет общая первая часть (3 первые цифры контекста), т.е. можно контексты все не перечислять, а указать общую часть контекста и по ней идентифицировать предприятие.

Что проще реализовать?

**Comment by prog-it** [comment]
Более правильно, конечно, использовать доп. таблицу в базе, но первый вариант проще и с вариантом 2 много нужно переписывать. Скорее всего, первый вариант и будет добавлен.
А насчет того, что "Астер сам пишет в базу", можно сделать через триггер в базе. Т.е. будет еще одна таблица в базе по варианту 2 и триггер на событие "insert". Триггер выполнит соответствие контекстов и вставит нужное имя предприятия в cdr. Но триггер, понятно, нужно будет написать самим.

**Comment by dimm71** [comment]
Прикрутил CDR-Viewer к FreeSwith, но хочется большего.
Возможно ли сделать такую штуку...
Сейчас в FS используется несколько доменов и в MySQL, домен записывается в отдельный столбец к каждому звонку. Очень хочется, чтобы была возможность сортировки данных по этому столбцу. Это тогда решит вопрос мультидоменности в FS. Естественно, данная сортировка должна настраиваться в конфиге, чтобы пользователям небыло доступа для смены домена.
Возможно ли это реализовать на текущей версии или будет такая возможность в ближайшее время?

**Comment by andrewnord** [comment]
тоже интересна возможность фильтровать по определенному столбцу прямо из конфига

**Comment by prog-it** [comment]
**andrewnord**
Т.е. задавать в конфиге имя столбца и значение, по которому фильтровать. В веб-интерфейсе, чтобы сразу уже выводились отфильтрованные данные?

**dimm71**

> Очень хочется, чтобы была возможность сортировки данных по этому столбцу

Только тут получается не сортировка по домену, а поиск? Вам же нужно выводить записи только по одному домену. Т.е. задать в конфиге имя домена и выводить записи только по нему, так?

> Кроме этого, хотелось бы вывести в статистику дополнительные колонки, указав самостоятельно какие именно.

Только вывести, или с возможностью поиска по ним? Если только вывести, то немного проще.

**Comment by dimm71** [comment]
Да. Я хотел бы вбить нужный домен в конфиге и чтобы поля выводились только те, где в строках присутствует этот домен. 

В принципе можно и поиск сделать, но хотя бы для  начала их вывести.

И такой вопрос... Если использовать разные конфиги и будет установлен пароль, то ведь любой желающий изменив в браузере строку ввода может попасть в любой профиль и посмотреть статистику. Или я ошибаюсь?!

**Comment by prog-it** [comment]
Постараюсь в ближайшее время добавить.

Да. Но, для этого нужно знать имя конфига, желательно задавать какое-нибудь уникальное имя.

**Comment by dimm71** [comment]
Отлично. Буду ждать.

**Comment by andrewnord** [comment]
> andrewnord
> Т.е. задавать в конфиге имя столбца и значение, по которому фильтровать. В веб-интерфейсе, чтобы сразу уже выводились отфильтрованные данные?

да, именно так

> Да. Но, для этого нужно знать имя конфига, желательно задавать какое-нибудь уникальное имя.
> А базовый конфиг (запрос без параметров ?config=) можно вообще удалить.

а нельзя добавить в config имена пользователей для которых будет доступен этот конфиг? по аналогии как задаются 'admins' => array( );, добавить  еще 'users' => array( ); - все остальным доступ закрыт
